runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/grpcclient/GrpcGetWorkerMetadataStream.java

```diff
@@ -0,0 +1,171 @@
+      Consumer<WindmillEndpoints> serverMappingUpdater) {
+    GrpcGetWorkerMetadataStream getWorkerMetadataStream =
+        new GrpcGetWorkerMetadataStream(
+            startGetWorkerMetadataRpcFn,
+            backoff,
+            streamObserverFactory,
+            streamRegistry,
+            logEveryNStreamFailures,
+            jobHeader,
+            metadataVersion,
+            getWorkerMetadataThrottleTimer,
+            serverMappingUpdater);
+    getWorkerMetadataStream.startStream();
+    return getWorkerMetadataStream;
+  }
+
+  @VisibleForTesting
+  static GrpcGetWorkerMetadataStream forTesting(
+      Function<StreamObserver<WorkerMetadataResponse>, StreamObserver<WorkerMetadataRequest>>
+          getWorkerMetadataRpc,
+      int metadataVersion,
+      JobHeader jobHeader,
+      Set<AbstractWindmillStream<?, ?>> streamRegistry,
+      Consumer<WindmillEndpoints> serverMappingUpdater) {
+    GrpcGetWorkerMetadataStream testStream =
+        new GrpcGetWorkerMetadataStream(
+            getWorkerMetadataRpc,
+            FluentBackoff.DEFAULT.backoff(),
+            StreamObserverFactory.direct(DEFAULT_STREAM_RPC_DEADLINE_SECONDS * 2, 1),
+            streamRegistry,
+            1, // logEveryNStreamFailures
+            jobHeader,
+            metadataVersion,
+            new ThrottleTimer(),
+            serverMappingUpdater);
+    testStream.startStream();
+    return testStream;
+  }
+
+  /**
+   * Each instance of {@link AbstractWindmillStream} owns its own responseObserver that calls
+   * onResponse(). This method does not and should not ever have any concurrent callers.
+   */
+  @Override
+  protected void onResponse(WorkerMetadataResponse response) {
+    if (response.getMetadataVersion() > metadataVersion) {
+      metadataVersion = response.getMetadataVersion();
+      serverMappingConsumer.accept(WindmillEndpoints.from(response));
+    } else {
+      // If the currentMetadataVersion is greater than or equal to one in the response, the response
+      // data is stale, and we do not want to do anything.
+      LOG.info(
+          "Received metadata version={}; Current metadata version={}. "
+              + "Skipping update because received stale metadata",
+          response.getMetadataVersion(),
+          metadataVersion);
+    }
+
+    this.latestResponse = response;
+  }
+
+  @Override
+  protected synchronized void onNewStream() {
+    send(workerMetadataRequest);
+  }
+
+  @Override
+  protected boolean hasPendingRequests() {
+    return false;
+  }
+
+  @Override
+  protected void startThrottleTimer() {
+    getWorkerMetadataThrottleTimer.start();
+  }
+
+  @Override
+  protected void sendHealthCheck() {
+    send(HEALTH_CHECK_REQUEST);
+  }
+
+  @Override
+  protected void appendSpecificHtml(PrintWriter writer) {
+    writer.format(
+        "GetWorkerMetadataStream: version=[%d] , job_header=[%s], latest_response=[%s]",
+        metadataVersion, workerMetadataRequest.getHeader(), latestResponse);
```

Comment:
I think you need some synchronization for latestResponse etc since I believe the html rendering can be concurrent

Reply:
done. added locking